File: LeetCode/Medium/0451-sort-characters-by-frequency/0451-sort-characters-by-frequency.py
class Solution:
    def frequencySort(self, s: str) -> str:
        count = [0] * 128
        freq = [0] * (len(s) + 1)
        ret = [""] * len(s)
        
        for v in s:
            temp = ord(v)
            count[temp] += 1
        
        for v in count:
            freq[v] += 1
        
        i = len(freq) - 1
        idx = 0
        
        while i > 0:
            while freq[i]:
                for j in range(128):
                    if count[j] == i:
                        char = chr(j)
                        for _ in range(i):
                            ret[idx] = char
                            idx += 1
                        
                        freq[i] -= 1
            i -= 1
            
        return "".join(ret)
                

File: LeetCode/Medium/0451-sort-characters-by-frequency/0451-sort-characters-by-frequency.cpp
class Solution {
public:
    string frequencySort(string s) {
        vector<int> count(128);
        vector<int> freq(s.size() + 1);
        vector<char> ret(s.size());
        
        for (char v : s)
            ++count[v];
        
        for (int v : count)
            ++freq[v];
        
        int i = freq.size() - 1;
        int idx = 0;
        
        while (i > 0){
            while (freq[i] > 0){
                for (int j = 0; j < 128; ++j){
                    if (count[j] == i){
                        for (int k = 0; k < i; ++k){
                            ret[idx++] = (char) j;
                        }

                        freq[i]--;
                    }
                }
            }
            i--;
            
        }

        return string(ret.begin(), ret.end());
    }
};